### PR TITLE
[ROCm] Improve hip-clang support in build_amd.py

### PR DESCRIPTION
**Pull request by iotamudelta**
Use the supported way to differentiate and automatically switch between hip-clang and hcc hipification in build_amd.py.

Cleaned up from PR #23699

### CODE REVIEW

File: tools/amd_build/build_amd.py [hunk 1]
```diff
@@ -142,10 +136,18 @@
                     f.flush()
                     os.fsync(f)
 
+# Check if the compiler is hip-clang.
+def is_hip_clang():
+    try:
+        hip_path = os.getenv('HIP_PATH', '/opt/rocm/hip')
+        return 'HIP_COMPILER=clang' in open(hip_path + '/lib/.hipInfo').read()
+    except Exception:
```
Comment (vincentqb) [inline tools/amd_build/build_amd.py]: What exceptions are you expecting? Is it IOError?

**Review by vincentqb — COMMENTED** [review]
Do you have documentation for "supported way to differentiate" ?

**Comment by iotamudelta** [comment]
@vincentqb no, we do not have documentation for that. I guess my statement is the documentation. :-)

File: tools/amd_build/build_amd.py (hunk 1 above)
Comment (iotamudelta) [reply tools/amd_build/build_amd.py]: Yes, that'd be the standard case - do you want me to specialize?

Comment (ezyang) [reply tools/amd_build/build_amd.py]: yes please

**Review by ezyang — APPROVED** [review]

**Pull request closed without merging**